### PR TITLE
Add support for variadic arguments on provider constructor

### PR DESCRIPTION
**Pull request by mumia**
Example:

```
dependencies.graph.Define(
        &dependencies.RouteRegistry,
        inject.NewProvider(
            routes.NewRouteRegistry,
            &dependencies.AuthenticationRoutes,
            &dependencies.AccountRoutes,
            &dependencies.BankRoutes,
            &dependencies.BaseRoutes,
        ),
    )
```
Where routes.NewRouteRegistry is:
```
func NewRouteRegistry(routes ...Route) RouteRegistry
```

### CODE REVIEW

**Comment by karlkfi** [comment]
Sorry for the long silence. 
I know I don't have CI or unit tests for every feature, but can you add a test that covers this feature and show that it and the other tests still pass?

**Comment by mumia** [comment]
@karlkfi We are already using this in a production environment. :) I will write some tests so you can merge it.
Thanks for the reply.

**Comment by mumia** [comment]
The tests are passing:
$ go test -v ./...
?   	github.com/mumia/inject	[no test files]
=== RUN   TestExtractByType
--- PASS: TestExtractByType (0.00s)
=== RUN   TestExtractByTypeNoMatch
--- PASS: TestExtractByTypeNoMatch (0.00s)
=== RUN   TestExtractByTypeMultiMatch
--- PASS: TestExtractByTypeMultiMatch (0.00s)
=== RUN   TestExtractAssignable
--- PASS: TestExtractAssignable (0.00s)
=== RUN   TestExtractAssignableNoMatch
--- PASS: TestExtractAssignableNoMatch (0.00s)
=== RUN   TestExtractAssignableMultiMatch
--- PASS: TestExtractAssignableMultiMatch (0.00s)
=== RUN   TestFindByType
--- PASS: TestFindByType (0.00s)
=== RUN   TestFindAssignable
--- PASS: TestFindAssignable (0.00s)
=== RUN   TestGraphSupportsInterfaces
--- PASS: TestGraphSupportsInterfaces (0.00s)
=== RUN   TestGraphSupportsStructPointers
--- PASS: TestGraphSupportsStructPointers (0.00s)
=== RUN   TestGraphSupportsProviderConstructorArgs
--- PASS: TestGraphSupportsProviderConstructorArgs (0.00s)
=== RUN   TestGraphSupportsAutoProvider
--- PASS: TestGraphSupportsAutoProvider (0.00s)
=== RUN   TestGraphSupportsDownCasting
--- PASS: TestGraphSupportsDownCasting (0.00s)
=== RUN   TestGraphSupportsPartialResolution
--- PASS: TestGraphSupportsPartialResolution (0.00s)
=== RUN   TestGraphLifecycle
--- PASS: TestGraphLifecycle (0.00s)
=== RUN   TestVariadicNone
--- PASS: TestVariadicNone (0.00s)
=== RUN   TestVariadicOne
--- PASS: TestVariadicOne (0.00s)
=== RUN   TestVariadicAll
--- PASS: TestVariadicAll (0.00s)
PASS
ok  	github.com/mumia/inject/test	0.009s
?   	github.com/mumia/inject/test/variadic	[no test files]

:)

File: test/variadic_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package test
+
+import (
+	"testing"
+
+	"github.com/karlkfi/inject"
+	"github.com/mumia/inject/test/variadic"
```
Comment (karlkfi) [inline test/variadic_test.go]: s/mumia/karlkfi/

**Comment by karlkfi** [comment]
Tests are failing because one of your imports is importing your fork, not the real path

File: test/variadic_test.go (hunk 1 above)
Comment (andrzejsliwa) [inline test/variadic_test.go]: @mumia please rename it, this will let it merge

Comment (mumia) [inline test/variadic_test.go]: @karlkfi sorry for the delay, fixed now

**Comment by karlkfi** [comment]
Thanks!

**Comment by hori-ryota** [comment]
This change has broken for slice arguments.
```
		if inType.Kind() == reflect.Slice {
			inType = inType.Elem()
		}
```

If arg type is `[]uint` , `inType` will be `uint` and fail `AssignableTo` and `ConvertibleTo` .

**Pull request merged**